### PR TITLE
feat(vscode-extension) error reporting, qol

### PR DESCRIPTION
**Pull request by alii**
### What does this PR do?

A few goodies implemented into the VSCode extension.
- A basic test runner that uses the TypeScript AST for finding tests. Most of this came from #13668
- Bring in the LifecycleAgent & co from #15194 
- Show errors inside of VSCode as they happen at runtime
- Implements a new DebugAdapter for communicating via a `node:net` Socket

### How did you verify your code works?

Ran extension and tested each feature

### Notes:

- Async stack traces are NOT supported. This is because JSC does not actually record them unless the debugger is open. The debugger is big overhead, so it's not open in this case.
- Ideally we would show the function name of `.relatedInformation` for each diagnostic. Right now we just show the error message again. This isn't _amazing_...
- Disabling this right now would be done by setting the `BUN_INSPECT_NOTIFY` variable to `''`. A use case for disabling would be like when running a benchmark file (where this would actually have a very very slight performance overhead)
- The socket framer on the extension side could be rewritten

### CODE REVIEW

File: packages/bun-vscode/package.json [hunk 1]
```diff
@@ -328,5 +357,8 @@
         }
       }
     ]
+  },
+  "dependencies": {
+    "strip-ansi": "^7.1.0"
```
Comment (Electroid) [inline packages/bun-vscode/package.json]: Let's remove this dependency, we don't need it. Can be 1 line utility somewhere

Comment (alii) [reply packages/bun-vscode/package.json]: Sure, removed.

File: src/bun.js/bindings/InspectorLifecycleAgent.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    Ref<JSON::ArrayOf<String>> urls = JSON::ArrayOf<String>::create();
+    Ref<JSON::ArrayOf<int>> lineColumns = JSON::ArrayOf<int>::create();
+    Ref<JSON::ArrayOf<String>> sourceLines = JSON::ArrayOf<String>::create();
+
+    for (size_t i = 0; i < exception.stack.source_lines_len; i++) {
+        sourceLines->addItem(exception.stack.source_lines_ptr[i].toWTFString());
+    }
+
+    for (size_t i = 0; i < exception.stack.frames_len; i++) {
+        ZigStackFrame* frame = &exception.stack.frames_ptr[i];
```
Comment (dylan-conway) [inline src/bun.js/bindings/InspectorLifecycleAgent.cpp:109]: In the future, I think we can get the function name from `frame` here

Comment (alii) [reply src/bun.js/bindings/InspectorLifecycleAgent.cpp:109]: Interesting, we'd need to add this to the inspector event. Probably can't guarantee that we'll always have it (@Jarred-Sumner ?)

**Review by dylan-conway — APPROVED** [review]

**Comment by alii** [comment]
cc @Jarred-Sumner @dylan-conway 
CleanShot 2024-11-21 at 17 28 38@2x
Could this actually be relevant? Or expected that some/certain tests are failing here

**Comment by dylan-conway** [comment]
Yup, this is a new test failure, looking into it. I believe the cause is from https://github.com/oven-sh/bun/pull/15261/files#diff-36302f74ed5420c7bee6a5460115be29fd82720720381c0d2c8bb7c934e69591L382

File: bench/snippets/INTERNAL_DEBUGGER
Comment (Electroid) [inline bench/snippets/INTERNAL_DEBUGGER:1]: Remove empty file

File: packages/bun-debug-adapter-protocol/src/debugger/adapter.ts [hunk 3]
```diff
@@ -274,26 +290,24 @@ export class DebugAdapter extends EventEmitter<DebugAdapterEventMap> implements
...
   /**
-   * Gets the inspector url.
+   * Gets the inspector url. This is deprecated and exists for compat.
+   * @deprecated You should get the inspector directly, and if it's a WebSocketInspector you can access `.url` direclty.
    */
   get url(): string {
-    return this.#inspector.url;
-  }
+    // This code has been migrated from a time when the inspector was always a WebSocketInspector.
+    // So t
```
Comment (Electroid) [inline packages/bun-debug-adapter-protocol/src/debugger/adapter.ts]: Remove

File: packages/bun-debug-adapter-protocol/src/debugger/adapter.ts [hunk 4]
```diff
@@ -437,18 +451,36 @@ export class DebugAdapter extends EventEmitter<DebugAdapterEventMap> implements
...
+    if (request.enableDebugger !== false) {
+      this.send("Debugger.enable").catch(error => {
+        const { message } = unknownToError(error);
+        if (message !== "Debugger domain already enabled") {
+          throw error;
+        }
+      });
+    }
+
     this.send("Debugger.setAsyncStackTraceDepth", { depth: 200 });
```
Comment (Electroid) [inline packages/bun-debug-adapter-protocol/src/debugger/adapter.ts]: Won't this command fail is `Debugger.enable` is not run above? Should it be moved into the new `if` block?

**Comment by alii** [comment]
This code (or at least something similar iirc) was removed previously, I think because it was printing errors twice. Did you test this locally and if so does it seem to be working? I can try it out as well to test

File: packages/bun-debug-adapter-protocol/src/debugger/adapter.ts (hunk 4 above)
Comment (alii) [reply packages/bun-debug-adapter-protocol/src/debugger/adapter.ts]: Should be, but interestingly it hasn't been failing from my testing. Will move it for the sake of being more correct though. nice spot thanks

**Pull request merged**